### ISSUE TITLE
Left join without rechunk causes segfault

### ISSUE THREAD

**Issue by SimonSchneider** [pola-rs/polars#5573]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://crates.io/crates/polars) of Polars.

### Issue description

When left joining two lazy dataframes without rechunking the left one the app crashes with a segfault.

I have two dataframes
- the left one has been manipulated quite a lot with various `when/then` combinations etc.. 
- the right one is read directly and not manipulated
when joining these two with
```
let res = left
        .lazy()
        .join_builder()
        .with(right.lazy())
        .left_on(vec![col("layer-guid")])
        .right_on(vec![col("guid")])
        .how(JoinType::Left)
        .finish()
        .collect()?;
```
the app crashes with 
```
[1]    33274 segmentation fault  cargo run --release -- ...args...
```
When rechunking the left DF it works as expected.

I found this when trying to create a reproducable example and not being able to get the test to fail after saving the DFs to parquet.

### Reproducible example

I've not been able to create a minimally reproducable example due to all the previous mutations being important in making it fail. But all of the mutations done are done with `when/then` chains.

### Expected behavior

not segfaulting, I guess rechunking internally if needed.
However we have a lot of cases where it works exactly like in this one but does not crash. So only specific cases or data cause the need to rechunk manually

### Installed versions

<details>
`["lazy", "is_in", "parquet", "json", "strings", "sort_multiple", "round_series", "dtype-categorical", "performant", "serde", "concat_str", "row_hash"]`
</details>

**Comment by ritchie46** [comment]
Could you run this with a debug build in miri or valgrind? Hard to know what to fix without an MWE.

**Comment by SimonSchneider** [comment]
@ritchie46 
Forgot to mention it but I tried that too but ran into the issue of stack overflow when trying the debug build 
https://github.com/pola-rs/polars/issues/4827

I understand it's quite little to go on debuggingwise... I thought I'd take a shot in the dark and at least report with as much info as I could produce, maybe you'd realise something. But for now I have a workaround which I don't think is too bad. Performance impact of always rechunking before joining shouldn't be too bad I guess.

Thanks for all the hard work on Polars, in GitHub issues one just comes to report issues, not so much give credit about an amazing lib. But know that it's incredible work you and the others are doing on Polars. Also stability has been improving lots since we first started using it back in Feb.

**Comment by ritchie46** [comment]
Well, that could be the segfault. Almost all segfaults mentioned in the issues occur due to a stackoverflow in rayon work strealing. Could you confirm that is the case?

> Thanks for all the hard work on Polars, in GitHub issues one just comes to report issues, not so much give credit about an amazing lib. But know that it's incredible work you and the others are doing on Polars. Also stability has been improving lots since we first started using it back in Feb.

Thanks a lot! I hope to start a foundation and get some sponsorships to improve the stability/ documentation, etc. much more.

**Comment by SimonSchneider** [comment]
Ah, I didn't see any mentions of segfaults in that issue, only the stack overflow in the debug build. Not sure how I'd confirm this though? 
Also the join issue happens after the any `whenThen` chains have already been collected to _non_ lazy frames. I also was suprised that the rechunk solved it but now that you say it could be that I'm guessing non rechunked creates work per chunk for rayon and could that work trigger the segfault?

> Thanks a lot! I hope to start a foundation and get some sponsorships to improve the stability/ documentation, etc. much more.
Nice

**Comment by ritchie46** [comment]
So you join on `DataFrame` not on `LazyFrame`? Could you add some pseudocode maybe, and given some information on your dtypes and amount of chunks?

**Comment by ritchie46** [comment]
@SimonSchneider is #5579 related?

**Comment by SimonSchneider** [comment]
maybe, I'll try out the new commit.

for reference this is what I did to get the behaviour, in pseudo code
```
fn runMain(inputDf: DataFrame, whenThens: WhenThenType[], otherDf: DataFrame) {
  // the following crashes with SO on debug build so I can't get the DataFrame into the broken state
  let modifiedDf: DataFrame = whenThens.reduce(inputDf, |df, wt| df.lazy().whenThen(wt).collect()?)?; 

  modifiedDf = modifiedDf.rechunk() // <- this line fixes the Segfault
  
  // the following crashes with Segfault on release build
  let outputDf: DataFrame = modifiedDf.lazy()
        .join_builder()
        .with(otherDf.lazy())
        .left_on(vec![col("layer-guid")])
        .right_on(vec![col("guid")])
        .how(JoinType::Left)
        .finish()
        .collect()?;
  return outputDf;
}
```

I'll come back with regarding if the new version fixes this.

**Comment by SimonSchneider** [comment]
Tried out the new commit and that did unfortunately not fix this

**Comment by ritchie46** [comment]
Can you share the schemas and the amount of chunks of the dataframes? Can you also check the sortedness flags? That should be enough to recreate an MWE.

And if you run with `POLARS_VERBOSE=1`, what is the output shown?

**Comment by SimonSchneider** [comment]
I'm not sure how to get the sortedness flags. But here's what I can gather. It also keeps getting weirder because the left dataframe reports that it has only 1 `chunk`, and the right one has 3 `chunks`. But when I rechunk the left one before joining it all works. The only debug log from `POLARS_VERBOSE=1` is `join parallel: true` when joining the DFs.

```
info:
 l-df: chunks: 1, sortedness: ?
 r-df: chunks: 3, sortedness: ?
 left-schema:
   u32 - __id1
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   i64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - datasetId
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   i64 - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - guid
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   f64 - redacted
   i64 - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   f64 - redacted
   str - redacted
   str - redacted
   f64 - redacted
   f64 - redacted
   str - redacted
   str - redacted
   str - plant::dataset::name
   str - BuildingElementName
   str - ProductBSAB
   str - BIMLayer
   f64 - ProductVolume
   f64 - ProductWeight
   f64 - ProductArea
   str - BIMEntity
   str - BIMType
   str - BIMName
   str - ProductName
   f64 - ProductLength
   f64 - ProductCost
   f64 - ProductThickness
   str - ProductUnit
   f64 - ProductUnitWeight
   str - BuildingElementUnit
   str - BuildingElementFileId
   f64 - BuildingElementArea
   str - BuildingElementComponentSubGroup
   str - BuildingElementRecycleType
   str - ProductRawMaterialId
   str - ProductRecycleType
   str - ProductComment
   f64 - ProductAmount
   f64 - BuildingElementAmount
   str - A5CalculationMethod
   str - plant::buildingElement::id
   bool - plant::buildingElement::merge
   str - plant::buildingElement::unit
   str - plant::buildingElement::name
   f64 - plant::buildingElement::amount
   str - layer-guid
 right-schema:
   str - redacted
   str - redacted
   str - redacted
   str - redacted
   str - guid
   f64 - redacted
   str - redacted
   str - redacted
joining
 Left: col("layer-guid") - 
 Right: col("guid") - 
join parallel: true
[1]    45366 segmentation fault  POLARS_VERBOSE=1 cargo run --release 
```

**Comment by SimonSchneider** [comment]
Thanks, we are getting closer I think, it seems some of the series have 1 chunk and others have 5, selecting one with 1 chunk and one with 5 seems to break it

```

info:
  left - (14195 x 2) - chunks: 1  <--- DF info
   dt: str - sort: non - chunks: 1 - len: 14195 - ProductName
   dt: str - sort: non - chunks: 5 - len: 14195 - layer-guid

  right - (26278 x 1) - chunks: 3 <--- DF info
   dt: str - sort: non - chunks: 3 - len: 26278 - guid
joining
 Left: col("layer-guid") - 
 Right: col("guid") - 
join parallel: true
[1]    52904 segmentation fault  POLARS_VERBOSE=1 cargo run --release
```
one note is that we are doing the same for many different datasets and it only happens to one of them which is very odd

**Comment by ritchie46** [comment]
Alright.. I think I know what might be it. I will come back to you.

**Comment by ritchie46** [comment]
Could you try if this fixes it? #5589

**Comment by SimonSchneider** [comment]
Yep that fixes, it! 👑

**Comment by ritchie46** [comment]
Great!

**Comment by ritchie46** [comment]
Closed by #5589

**Comment by SimonSchneider** [comment]
Thanks a bunch!
